### ISSUE TITLE
Docs for R6 stubbing

### ISSUE THREAD

**Issue by kenahoo** [jfiksel/mockery#21]
I see that there's a closed ticket (#13) for mocking R6 objects and their methods.  I'm not quite sure how to use this, though, could some docs be added?

Specifically, when I call function `func` below, I'd like to mock the class `MyClass` so that an object of that class has no-op methods automatically stubbed, and I can inspect (after the fact) when they're called.

```
func <- function(...various.args.here...) {
  obj <- MyClass$new()
  obj$my_method(...)
  obj$my_other_method(...)
  # ... lots of calls to various other `obj` methods
}
```

Is this a supported use case?  Would it automatically also override method calls on that object that happen in a function *called* from `func` (or is that maybe what the `depth` argument controls)?

Thanks!

**Comment by csdesml** [comment]
No, this is not a supported use case. If you're creating the object in the function instead of passing it in, we cannot then pull it out (unless you stub the `MyClass$new()` method in `func`'s scope to return some mock class, which is doable but I think not what you're asking).

What you can do is pass `obj` in as an argument, in which case this becomes doable.

```
obj = MyClass$new()
obj$my_method <- mockery::mock()
func(...other args, obj)
mockery::expect_called(obj$my_method, 'expected argument')
```

Please see the read me for further documentation about usage of mocks and stubs.  In general dependency injection makes testing a lot easier.

Does that answer your question?

**Comment by kenahoo** [comment]
Thanks for the response.  I'm flexible about how I approach this, just trying to figure out what the options are.

Stubbing `MyClass$new` was one of the options I tried, but I couldn't make the pieces fit together successfully.

I'm experimenting with something similar to your suggestion now - I added an argument to `func`, defaulting to `MyClass`, that gives the class (or whatever R6 calls its object-creation-objects) that should be instantiated.  That facilitates testing, but I'm not crazy about having to add an additional argument to the tested code.
